### PR TITLE
Added ModelIllegalStateException to handle lazy loading exception.

### LINKED ISSUE #9645: Handling lazy loading exceptions for storage in new and old storage
### Describe the bug

When lazy-loading information from the JPA Map storage, the test ConcurrencyTest#createRemoveClient() showed NPEs when a clients was removed while populating the list of clients. 

With the old JPA storage this was handled that all exceptions during lazy loading where caught and the entities involved would be removed from the list of clients returned to the caller using `try { c.getClientId(); return true; } catch (Exception ex) { return false; }`.

### Version

17.0.0-SNAPSHOT

### Expected behavior

* Throw a specific exception related to lazy loading across all storage implementations
* Don't depend on a storage provider specific specific field as with`c.getClientId()` to trigger the exception
* Document the behavior of removing clients from the client list when a lazy loading exception occurs
* Handle lazy loading situations in map JPA storage implementation for deletion and modification of entities

### Actual behavior

* All exceptions being caught
* Provider specific field to trigger lazy loading problems
* No documentation present (neither for the caller, nor in the code)
* No handling of lazy loading issues in JPA map storage provider

### How to Reproduce?

Run ConcurrencyTest#createRemoveClient() with map-jpa storage configuration.

### Anything else?

There's been a GitHub discussion to find out what needed to be done: https://github.com/keycloak/keycloak/discussions/9601
This will be applied in all places where the pattern of filtering out clients is used.

Also: 
* Implement similar handling for Map JPA Roles.
* Revisit https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L157-L166 to see if the isolation warning can now be dropped.

Out of scope: 
* Client will not be able to identify what clients have been removed - this is the same behavior we have today
* All illegal state exceptions will be handled the same way, both removed and modified clients will lead to the clients not being listed in the caller's result.
* No retrying of the request or return a 500 error, as this could potentially lead to a denial of service situation when a lot of clients are modified concurrently.

### PR DESCRIPTION
**Pull request by ahus1**
Closes #9645

### CODE REVIEW

**Comment by ahus1** [comment]
This makes sure that the updated delegate exists in the database and that its optimistic locking version hasn't changed. 

After this change the ConcurrencyTest with JPA Map storage completes. 

Design changes: 
* common base class for all JpaDelegateProviders to write the code only once. Pushing up the delegate instance variable. To enable the comparison interface JpaRootEntity introduced
* ClientService will filter out invalid clients. 
* Interface didn't match well for logoutAll(), therefore therefore catching the exception here
* ExportUtils will fail if a client can't be exported ... this changes the existing behavior, but seems to be reasonable as exportutils should export everything, and no clients shuold be missing.

Optional stuff (please comment):
* Entity manager could be pushed to JpaDelegateProvider as well.

**Comment by ahus1** [comment]
Testing this yesterday, I found that the optimistic locking failed when updating only some attributes: the version in the parent wasn't incremented as the the ManyToOne relation is owned by the child.

I looked for a simple solution to change that ownership, but didn't succeed. I then added some hibernate listeners and iterfaces to trigger the essential `session.lock(root, LockModeType.OPTIMISTIC_FORCE_INCREMENT)`. 

I added a second commit to this PR to keep them separate for an initial review and comment. After initial feedback on the approach I'll then merge them. 

The approach is inspired by https://vladmihalcea.com/jpa-entity-version-property-hibernate/ and https://vladmihalcea.com/hibernate-event-listeners/.

I tested it manually by changing role attribute or changing a client's "Backchannel Logout Session Required"

@vramik, @sguilhen, @hmlnarik - I'd be happy if you could have a look.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java [hunk 1]
```diff
@@ -583,7 +586,6 @@ public void removeAttribute(String name) {
             JpaClientAttributeEntity attr = iterator.next();
             if (Objects.equals(attr.getName(), name)) {
                 iterator.remove();
-                attr.setClient(null);
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:586]: Could you please explain why this change is necessary? AFAIUndersood it is required for syncing when using bidirectonal `@OneToMany` where `@ManyToOne` side manages the association.

**Review by vramik — COMMENTED** [review]
Thank you @ahus1. On the first look it seems good to me. I haven't tried it yet. 

I've noticed one thing (please see inline) what attracted my attention. Could you please in the meantime take a look while I'll take a closer look. Thank you.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java (hunk 1 above)
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:586]: I found that the `@PreDelete` handler needed to know the parent entity to enforce the locking. When setting it to null, this wan't possible any more.

Then I digged deeper into the Hibernate things and found that "orphanRemoval" should do the job. Then I tested it, and the attribute were removed correctly. "orphanRemoval" is a JPA 2 feature. This might have been necessary in the past, but my manual tests showed that this is no longer necessary; it is sufficient to remove the item from the collection.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java:586]: I see, thank you for the explanation.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Interface for all root entities in the JPA storage.
+ */
+public interface JpaRootEntity {
+
+    /**
+     * Version of the JPA entity used for optimistic locking
+     */
+    int getVersion();
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java:27]: nitpick: could there be `@Override` annotation to implementations?

**Review by vramik — COMMENTED** [review]
I've looked mostly on jpa part and from my POV it looks good. I've added few minor comments.

Thank you @ahus1.

File: server-spi-private/src/main/java/org/keycloak/models/utils/ModelToRepresentation.java [hunk 3]
```diff
@@ -17,6 +17,8 @@
 
 package org.keycloak.models.utils;
 
+import org.jboss.logging.Logger;
+import org.keycloak.TokenVerifier;
```
[the remark this answers is not in the document]
Comment (ahus1) [reply server-spi-private/src/main/java/org/keycloak/models/utils/ModelToRepresentation.java:21]: thanks, will change it.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java (hunk 2 above)
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java:27]: +1!

**Comment by ahus1** [comment]
Thanks @vramik for the findings, I update the PR with once commit that contains all of your changes. To make it simpler for @sguilhen and @hmlnarik to review, I kept the commits separate. That should make it simpler to decide if both steps should be taken.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java [hunk 4]
```diff
@@ -107,7 +107,6 @@ public static RealmRepresentation exportRealm(KeycloakSession session, RealmMode
 
         if (options.isClientsIncluded()) {
             clients = realm.getClientsStream()
-              .filter(c -> { try { c.getClientId(); return true; } catch (Exception ex) { return false; } } )
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:110]: This filtering should be kept by using `ModelToRepresentation.filterValidRepresentations(realm.getClientsStream(), app -> exportClient(session, app)` below.

Comment (ahus1) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:110]: Hi @hmlnarik, I did this change intentionally as stated above, see my comment: 

> ExportUtils will fail if a client can't be exported ... this changes the existing behavior, but seems to be reasonable as exportutils should export everything, and no clients should be missing.

I now change the code as you requested. As the code needs a list of ClientModels later to process groups etc., the code is now a bit more complex to construct the list of client models that are valid client representations. Please let me know what you think.

**Comment by ahus1** [comment]
After a conversation with @hmlnarik the changes about the hibernate listener described in https://github.com/keycloak/keycloak/pull/9761#issuecomment-1021047711 are no longer part of this ticket; they will be covered in a future ticket. Work on that will start once this has been merged.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 4 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:110]: A client might be missing permanently, leading to permanently thrown exception, and throwing an exception effectively removes ability to run export; so I'd rather keep it as is with your current changes.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java [hunk 5]
```diff
@@ -103,15 +101,17 @@ public static RealmRepresentation exportRealm(KeycloakSession session, RealmMode
                 .map(ClientScopeModel::getName).collect(Collectors.toList()));
 
         // Clients
-        List<ClientModel> clients = Collections.emptyList();
+        List<ClientModel> clients = new ArrayList<>();
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:104]: Please either use `ArrayList(int initialCapacity)` or prefer `LinkedList` since the list is only added to and iterated.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 4 above)
Comment (ahus1) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:110]: I could see a situation where someone would be happy to know about this problem by seeing the exception, and I also see the case you make here. 

When you're happy with the change, then please resolve this thread.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 5 above)
Comment (ahus1) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java]: Changed it to a LinkedList as I don't want to ask a stream about its size. Pushed the code, please resolve if you're ok with it.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
```
Comment (sguilhen) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaRootEntity.java:2]: Nitpick: year

File: server-spi/src/main/java/org/keycloak/models/ModelIllegalStateException.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ */
+package org.keycloak.models;
+
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * Thrown when data can't be retrieved for the model.
+ *
+ * This occurs when an entity has been removed or update in the meantime. This might wrap an optimistic lock exception
```
Comment (sguilhen) [inline server-spi/src/main/java/org/keycloak/models/ModelIllegalStateException.java:25]:
```suggestion
 * This occurs when an entity has been removed or updated in the meantime. This might wrap an optimistic lock exception
```

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java [hunk 8]
```diff
@@ -71,11 +60,20 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:74]: @hmlnarik Do we have a template for the imports? Some files in the codebase have the java and javax packages first, others have them at the end of the imports list. Just wondering because IIRC we've been keeping java and javax at the top.

Comment (sguilhen) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:74]: Regardless of the case, we have always avoided introducing these import reorganizations in commits.

File: server-spi-private/src/main/java/org/keycloak/models/utils/ModelToRepresentation.java [hunk 9]
```diff
@@ -554,6 +558,25 @@ public static RealmEventsConfigRepresentation toEventsConfigReprensetation(Realm
...
+     *
+     * Entities that have been removed from the store or where a lazy loading exception occurs will not show up
+     * in the output stream.
+     */
+    public static <M, R> Stream<R> filterValidRepresentations(Stream<M> models, Function<M, R> transformer) {
+        return models.map(m -> {
+                    try {
+                        return transformer.apply(m);
+                    } catch (ModelIllegalStateException e) {
+                        LOG.warn("unable to retrieve client information for logout", e);
```
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/models/utils/ModelToRepresentation.java:573]: I'm confused about this message - the contexts in which this method are called don't seem to be related to logout event.

File: services/src/main/java/org/keycloak/services/resources/admin/ClientsResource.java [hunk 10]
```diff
@@ -87,9 +86,11 @@ public ClientsResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEve
     }
 
     /**
-     * Get clients belonging to the realm
+     * Get clients belonging to the realm.
      *
-     * Returns a list of clients belonging to the realm
+     * Returns a list of clients belonging to the realm. If a clients can't be retrieved from the storage due
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/services/resources/admin/ClientsResource.java:91]:
```suggestion
     * Returns a list of clients belonging to the realm. If a client can't be retrieved from the storage due
```

File: services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java [hunk 11]
```diff
@@ -241,15 +242,18 @@ protected Response sendBackChannelLogoutRequestToClientUri(ClientModel resource,
...
-            GlobalRequestResult currentResult = logoutClient(realm, r, realm.getNotBefore());
-            finalResult.addAll(currentResult);
+        realm.getClientsStream().forEach(c -> {
+            try {
+                counter.getAndIncrement();
+                GlobalRequestResult currentResult = logoutClient(realm, c, realm.getNotBefore());
+                finalResult.addAll(currentResult);
+            } catch (ModelIllegalStateException ex) {
+                // currently, GlobalRequestResult doesn't allow for information about clients that we were unable to retrieve.
+                logger.warn("unable to retrieve model information, skipping entity", ex);
```
Comment (sguilhen) [inline services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java:255]: Message again: now that we are in the context of a logout there's no mention of it in the log message. Perhaps it is unnecessary but to me the different messages are confusing.

**Review by sguilhen — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java (hunk 11 above)
Comment (ahus1) [reply services/src/main/java/org/keycloak/services/managers/ResourceAdminManager.java:255]: Thanks for spotting this. Apparently I mixed up the two log message you commented on, the next version will have them switched.

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 8 above)
Comment (ahus1) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:74]: I had the auto-format for imports enabled when I missed some unused imports the other day. Will revert this change, and re-calibrate my IDE. If there is some information how to configure IntelliJ, or some rules that I can translate for IntelliJ, I'd be happy to do so. Another rule I learned in another review: no star imports :roll_eyes:

**Comment by ahus1** [comment]
@sguilhen - Thanks for the review. I updated this PR according to your review comments. Please have another go if this is good to go. Thanks!

File: services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java (hunk 8 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/exportimport/util/ExportUtils.java:74]: I don't think there is any strict rule for import ordering.
Generally the aim is to keep PRs minimal and focused on the particular task so the aim is to prevent large purely syntactical changes.

**Review by hmlnarik — APPROVED** [review]
Thank you for the PR!
